Allow selector ordering to ensure theme.json root selector margin takes precedence

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->


## Description

Companion to Core patch https://core.trac.wordpress.org/ticket/54550.

(originally an alternative approach to https://github.com/WordPress/gutenberg/pull/36958).

Note that as with https://github.com/WordPress/gutenberg/pull/36958 this PR ~also includes changes from #36957 which make it possible for you to test this PR. They will be removed prior to merge~ requires you to test against WordPress **5.8** (that's five point **EIGHT** not NINE). See https://github.com/WordPress/gutenberg/pull/36957#issuecomment-981759580 for reasons.

To use WP 5.8 add this to your [`.wp-env.override.json` file](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-env/#wp-env-override-json):

```
{
    "core": "WordPress/WordPress#5.8.2"
}
```

Closes https://github.com/WordPress/gutenberg/issues/36147

## How has this been tested?
Follow same instructions as https://github.com/WordPress/gutenberg/pull/36958.

But this time you should see that the `body { margin: 0; }` declaration is still output _but_ is overridden by values from `theme.json` if they are found. 

Basically the default `margin: 0` is still applied but anything in `theme.json` will win out.

I believe this is the best approach because

1. It is limited in scope.
2. It retains the original `margin: 0` on `body` which Themes have come to expect thus avoiding backwards compat issues.
3. It caters for both types of `margin` notation (i.e. string or object) that can be applied in `theme.json`.

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

getdave commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -1059,7 +1076,7 @@ private function get_block_classes( $style_nodes ) {
 			}
 
 			if ( self::ROOT_BLOCK_SELECTOR === $selector ) {
-				$block_rules .= 'body { margin: 0; }';
```
Avoid applying rule here otherwise it will end up being applied last in the selector order thereby ensuring it takes precedence over rules applied via `theme.json`.

getdave commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -1049,6 +1049,23 @@ private function get_block_classes( $style_nodes ) {
+			// `margin` in its `spacing` declaration for the `body` element then these
+			// user-generated values take precedence in the CSS cascade.
+			// See: https://github.com/WordPress/gutenberg/issues/36147.
+			if ( self::ROOT_BLOCK_SELECTOR === $selector ) {
```
By applying the margin: 0 rule early we ensure that any values provided from `theme.json` will be later in the selector order and thus take precedence over the default.

getdave commented:
cc @overclokk as you raised the original Issue.

ntsekouras reviewed (commented):
Thanks for the PRs Dave! I tested and work as expected.

I think I prefer this PR from the conditional zeroing of margin - also in the other approach  we should also take into account the format of `margin:{top: '10px'}` etc.

I'm just wondering if it worths to call the functions again, instead of just
```
if ( self::ROOT_BLOCK_SELECTOR === $selector ) {
	$block_rules .= 'body { margin: 0; }';
}
```
before the generation of the styles.

getdave commented:
> I'm just wondering if it worths to call the functions again, instead of just
> 
> ```
> if ( self::ROOT_BLOCK_SELECTOR === $selector ) {
> 	$block_rules .= 'body { margin: 0; }';
> }
> ```
> 
> before the generation of the styles.

I did this because it was putting the selectors on the _same line_ within the outputted code so I thought perhaps it was safer to reuse the function.

overclokk commented:
I was thinking why not adding that value as default in theme.json in core, so the theme can override that vales without an if statement

getdave commented:
In light of https://github.com/WordPress/gutenberg/pull/36957#issuecomment-981759580 I've removed the fix I cherrypicked from https://github.com/WordPress/gutenberg/pull/36957.

You now need to test this PR against WordPress **5.8**.

getdave committed:
Set default margin on body prior to generating styles from theme.json

getdave commented:
> I was thinking why not adding that value as default in theme.json in core, so the theme can override that vales without an if statement

Responded in https://github.com/WordPress/gutenberg/pull/36958#discussion_r759074183.

ntsekouras commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -714,6 +714,23 @@ private function get_block_classes( $style_nodes ) {
+			// `margin` in its `spacing` declaration for the `body` element then these
+			// user-generated values take precedence in the CSS cascade.
+			// See: https://github.com/WordPress/gutenberg/issues/36147.
+			if ( self::ROOT_BLOCK_SELECTOR === $selector ) {
```
I think the comment should be more about why we are zeroing the margin (default browsers margin) and not about the specific issue.

getdave replied:
Good idea! I had no idea why we were zeroing the margin. I assumed it was a reset or normalization but thanks for clarifying.

I'll update the comment. I would like to retain the explanation about the ordering as it's important to avoid a regression.

getdave replied:
Fixed in https://github.com/WordPress/gutenberg/pull/36960/commits/f7ed1177cbf4034eeb03ec7fdeb60a968b06cbce

ntsekouras reviewed (approved):
Thanks Dave for your work here! IMO this fix can be landed and discuss afterwards a different approach that would potentially change where we apply all the root styles from `theme.json`, given the WP release phase we are right now.

I'd give it a bit more time before merging if anyone else wants to chime in - cc @youknowriad

getdave committed:
Add additional context to comment

Addresses https://github.com/WordPress/gutenberg/pull/36960#discussion_r759093111

getdave committed:
Restore simple version of body rule

Also avoids creating failing unit tests due to a single space character changing in a snaphshot!

youknowriad commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -724,7 +734,6 @@ private function get_block_classes( $style_nodes ) {
 			}
 
 			if ( self::ROOT_BLOCK_SELECTOR === $selector ) {
```
Makes me wonder if all the content of this condition should be above but it's good regardless.

getdave commented:
Ok then folks, it seems [the majority are agreed](https://github.com/WordPress/gutenberg/pull/36996) that _this_ PR is the approach we want to pursue.

I will close out https://github.com/WordPress/gutenberg/pull/36996 and then make a Core patch for the changes in this PR.

getdave commented:
Companion Core patch in https://core.trac.wordpress.org/ticket/54550 cc @ntsekouras

getdave committed:
Update with fixes made in Core patch

See https://github.com/WordPress/wordpress-develop/pull/1989

getdave committed:
Fix output now includes new line.

getdave committed:
Fix tests relating to spacing

getdave committed:
Fix test to put body CSS reset first

getdave commented:
Just fixing the PHP unit tests prior to merging.
